rust/connlib/tunnel/src/tests.rs

```diff
@@ -251,14 +292,43 @@ impl StateMachineTest for TunnelTest {
         state.advance(ref_state, &mut buffered_transmits);
 
         // Assert our properties: Check that our actual state is equivalent to our expectation (the reference state).
-        assert_eq!(
-            state.gateway_received_icmp_requests, ref_state.gateway_received_icmp_requests,
-            "Packets should flow from client to gateway"
-        );
-        assert_eq!(
-            state.client_received_icmp_replies, ref_state.client_received_icmp_replies,
-            "Packets should flow from gateway to client"
-        );
+        for (resource_dst, seq, identifier) in ref_state.expected_icmp_handshakes.iter() {
+            let client_sent_request = &state
+                .client_sent_icmp_requests
+                .get(&(*seq, *identifier))
+                .unwrap();
+            let client_received_reply = &state
+                .client_received_icmp_replies
+                .get(&(*seq, *identifier))
+                .unwrap();
+            let gateway_received_request = &state
+                .gateway_received_requests
+                .get(&(*seq, *identifier))
+                .unwrap();
+
+            assert_eq!(
+                gateway_received_request.source(),
+                ref_state
+                    .client
+                    .tunnel_ip(gateway_received_request.source()),
+                "ICMP request on gateway to originate from client"
+            );
```

Comment:
we might want to remove this? if I understad the `gateway_received_request` is what the gateway decapsulates and is going to insert into the tunnel device later on.

So for now that's correct, but in the future the gateway might mangle the source ip and then map it back

Reply:
But the source IP always has to be the client's TUN device, no?